### PR TITLE
feat(eslint-plugin): add space-before-function-paren [extension]

### LINKED ISSUE #707: `comma-spacing` and `space-before-function-paren` are conflict in type
Don't know what's the root cause of this problem, I guess it's the parser

**What code were you trying to parse?**
```ts
function x<R extends Record<string, (this: any) => any>>(fns: R) {
    return fns
}
```
when eslint config includes
```json
"comma-spacing": "warn",
"space-before-function-paren": ["warn", "never"],
```

**What did you expect to happen?**
code shall pass lint

**What actually happened?**
code violates rule `comma-spacing` when
```ts
function x<R extends Record<string,(this: any) => any>>(fns: R) {
    return fns
}
```
and violates `space-before-function-paren` when
```ts
function x<R extends Record<string, (this: any) => any>>(fns: R) {
    return fns
}
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `1.12.0` |
| `TypeScript`                | `3.5.3` |
| `ESLint`                    | `6.0.2` |

### PR DESCRIPTION
**Pull request by Austaras**
"Half fixes" #707

### CODE REVIEW

**Comment by Austaras** [comment]
@bradzacher may you or other member take some time review this and get it merged?

File: packages/eslint-plugin/src/rules/space-before-function-paren.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+} from '@typescript-eslint/experimental-utils';
+import { isOpeningParenToken } from 'eslint-utils';
+import * as util from '../util';
+
+type Option = 'never' | 'always';
+type FuncOption = Option | 'ignore';
+
+export type Options = [
+
+    | Option
```
[the remark this answers is not in the document]
Comment (Austaras) [reply packages/eslint-plugin/src/rules/space-before-function-paren.ts]: maybe you should update prettier in `yarn.lock` first or test for format will fail

**Comment by bradzacher** [comment]
still formatting errors in the file - if you can run prettier on the file to get the CI passing, we can look to merge this.

**Comment by Austaras** [comment]
It should. If I guess correctly, CI is using prettier 1.18.2 in yarn.lock but your requested change could only satisfy format for prettier 1.19

**Comment by bradzacher** [comment]
oh wow. I'm sorry - I didn't realise we hadn't updated the prettier version.
My mistake. Thanks for this.

**Pull request merged**